feat: Upgrade desktop GUI to Bootstrap 5

kevcodez opened the pull request:
Bootstrap 3 went end of life on 24th of july 2019. This PR upgrades Bootstrap and related CSS/SASS plugins.

Addresses #9012 

* Update all CSS/SASS related plugins
  * Bump css-loader from 2.x to 4.x
  * Bump mini-css-extract-plugin from 0.x to 1.x
  * Bump postcss-loader from 3.x to 4.x
  * Bump sass-loader
* Upgrade Bootstrap 3 to Bootstrap 5
* Use Bootstrap 5 compatible modal
* Remove unused Bootstrap sass elements such as input-groups, button-groups and labels

### User facing changelog

* Desktop GUI CSS Framework has been updated from Bootstrap 3 to Bootstrap 5
* Bumped css-loader, mini-css-extract-plugin, postcss-loader and sass-loader to latest versions in Desktop GUI

### PR Tasks
* [x] Have tests been added/updated?

kevcodez commented on packages/desktop-gui/cypress/integration/footer_spec.js:
```diff
@@ -82,7 +82,7 @@ describe('Footer', () => {
       cy.get('.version').click()
       cy.get('.update-modal .close').click()
-      cy.get('.update-modal').should('not.exist')
+      cy.get('.update-modal').should('not.be.visible')
```
The new bootstrap4 dialog is still in the DOM but not visible, thus the change.

kevcodez commented on packages/desktop-gui/src/app/nav.scss:
```diff
@@ -4,6 +4,7 @@
   flex-grow: 0;
   flex-shrink: 0;
   flex-wrap: wrap;
+  padding: 0;
```
bootstrap 4 navbar has default padding that we need to get rid of

kevcodez commented on packages/desktop-gui/src/project-nav/browsers_spec.jsx:
```diff
@@ -59,7 +59,9 @@ describe('Browsers', () => {
     cy.log('**closing**')
-    cy.get('.spacer').click()
+    // Click anywhere next to the dropdown to make sure it closes
+    cy.get('.spacer').click({ force: true })
```
Without the force, the element was not visible as it's an "invisible" spacer - not sure why I had to adjust this as the previous test worked

kevcodez commented on packages/desktop-gui/src/project-nav/browsers_spec.jsx:
```diff
@@ -59,7 +59,9 @@ describe('Browsers', () => {
-    cy.get('.spacer').click()
+    // Click anywhere next to the dropdown to make sure it closes
+    cy.get('.spacer').click({ force: true })
+    browserListClosed()
```
Added the assertion to make sure the browser list actually closed after clicking somewhere else

kevcodez commented on packages/desktop-gui/src/styles/variables.scss:
```diff
@@ -34,7 +34,7 @@ $screen-md: 1200px; // Medium screen / desktop
 
 //** Padding applied to the modal body
-$modal-inner-padding:         15px !default;
+$modal-inner-padding:         0.9375rem !default;
```
Bootstrap 4 changed the unit from px to rem. We need to use the same unit as bootstrap because bootstraps calculation will fail when mixing px and rem

kevcodez commented:
I don't really get why the 4 tests fail. Locally with run only two tests fail. Locally when I open cypress none fails. Appreciate some help.

jennifer-shehane commented:
@kevcodez May be some variability in the speed the test runs. You may want to try running in Electron locally (since that's what's run in CI). I wish we could see the Cypress Dashboard results, but we can't record on outside contributor PRs. :(

kevcodez commented:
@jennifer-shehane Great tip, I was able to reproduce. Found the issue and fixed it.

jennifer-shehane reviewed (changes requested):
I want to get this PR in first, so that we can have a baseline of percy snapshots to compare the new CSS changes to. https://github.com/cypress-io/cypress/pull/9087

But, just looking preliminarily (I quickly pulled it down, didn't look at everything) there appears to be some unintended CSS changes to the UI. 

<img width="780" alt="Screen Shot 2020-11-04 at 3 01 31 PM copy" src="https://user-images.githubusercontent.com/1271364/98087874-22fc4400-1eaf-11eb-90c7-e7147fb57342.png">
<img width="780" alt="Screen Shot 2020-11-04 at 3 01 58 PM copy" src="https://user-images.githubusercontent.com/1271364/98087881-25f73480-1eaf-11eb-83ac-12e662553a18.png">

kevcodez commented:
@jennifer-shehane Fixed the margin on the intro view, button styles and the duplicate arrow

kevcodez commented:
@jennifer-shehane Presumably fixed browser icon alignment and fixed specs run button styling

kevcodez commented:
@jennifer-shehane By the way, is there a way to run the production bundle locally so I can actually login with my dashboard account and test the production application?

I know that I can build the production bundle and I also know that I can't login with development build because it only works for employees. Can I bundle it in a way that I can login with my account locally?

jennifer-shehane commented:
That's a good question, I'll have to ask. 

One of the best ways to test the views is through our Desktop-gui Cypress tests though, you should be able to see all of the views through there so you wouldn't need to log in. Running `yarn workspace @packages/desktop-gui cypress:open` and choosing the test with the view you want to tweak the styles of. 

More instructions for running the `desktop-gui` tests can always be found here: https://github.com/cypress-io/cypress/blob/develop/packages/desktop-gui/README.md

kevcodez commented:
Alright, that's why I am doing, too. Were will I be able to see the screenshot comparison of the develop vs. PR build?

When I get an overview of the unintended changes, I'll fix them

jennifer-shehane commented:
Apparently our Percy snapshots don't run in CI with external contributors since the env vars are unavailable 😑

kevcodez commented:
Can I somehow run that locally to compare?

kevcodez commented:
@jennifer-shehane What's the best way to continue with this?

jennifer-shehane reviewed (changes requested):
Sorry for the delay. We were getting 6.0 out. 

I ran Percy compare locally. Unfortunately this will not be available to contributors since our API key is required to run. I think you can view the run here: https://percy.io/cypress-io/cypress/builds/8019993?utm_campaign=cypress-io&utm_content=cypress&utm_source=github_status_public

There are still considerable differences to the UI with the new Bootstrap layout that would need to be addressed to accept this PR. 

Our team does not have time to spend on these fixes, as we have other priorities in our Roadmap that take precedent over this. But we would be willing to review the UI fixes.

**Some notable changes that are problematic**
- Some buttons are showing as gray that were previously white. 
- The text is a lot larger almost everywhere, leading some UIs to not display correctly due to overflow. 
- The modals are not taking up the desired width, so they look smushed. 
- The `cypress run` command instructions are showing up with a white background making it unreadable.

**Example screenshots before/after**

<img width="1526" alt="Screen Shot 2020-12-09 at 2 47 25 PM" src="https://user-images.githubusercontent.com/1271364/101603908-a4a33c80-3a2e-11eb-8a0c-45e9aa129eaf.png">
<img width="1518" alt="Screen Shot 2020-12-09 at 2 51 36 PM" src="https://user-images.githubusercontent.com/1271364/101603923-a836c380-3a2e-11eb-9892-43f8248409a1.png">
<img width="1522" alt="Screen Shot 2020-12-09 at 2 51 21 PM" src="https://user-images.githubusercontent.com/1271364/101603927-a967f080-3a2e-11eb-8eb0-eb4aeff29bad.png">
<img width="1518" alt="Screen Shot 2020-12-09 at 2 51 11 PM" src="https://user-images.githubusercontent.com/1271364/101603930-ab31b400-3a2e-11eb-9d64-b87f140bee8e.png">
<img width="1519" alt="Screen Shot 2020-12-09 at 2 51 05 PM" src="https://user-images.githubusercontent.com/1271364/101603934-abca4a80-3a2e-11eb-8cb9-55992215e85e.png">
<img width="1517" alt="Screen Shot 2020-12-09 at 2 50 58 PM" src="https://user-images.githubusercontent.com/1271364/101603937-ac62e100-3a2e-11eb-993b-a332f4f6feee.png">
<img width="1520" alt="Screen Shot 2020-12-09 at 2 50 43 PM" src="https://user-images.githubusercontent.com/1271364/101603941-acfb7780-3a2e-11eb-939f-46cd880c3965.png">
<img width="1518" alt="Screen Shot 2020-12-09 at 2 50 29 PM" src="https://user-images.githubusercontent.com/1271364/101603943-ad940e00-3a2e-11eb-9c24-b6a68e2ffca6.png">
<img width="1522" alt="Screen Shot 2020-12-09 at 2 50 17 PM" src="https://user-images.githubusercontent.com/1271364/101603948-ad940e00-3a2e-11eb-8198-84612455404a.png">
<img width="1519" alt="Screen Shot 2020-12-09 at 2 49 57 PM" src="https://user-images.githubusercontent.com/1271364/101603949-ae2ca480-3a2e-11eb-9fb1-bd2b25544362.png">
<img width="1523" alt="Screen Shot 2020-12-09 at 2 49 42 PM" src="https://user-images.githubusercontent.com/1271364/101603952-aec53b00-3a2e-11eb-88de-a10430c3bf92.png">
<img width="1522" alt="Screen Shot 2020-12-09 at 2 49 30 PM" src="https://user-images.githubusercontent.com/1271364/101603954-af5dd180-3a2e-11eb-93dd-fd6fecb6b6ec.png">
<img width="1520" alt="Screen Shot 2020-12-09 at 2 49 11 PM" src="https://user-images.githubusercontent.com/1271364/101603956-aff66800-3a2e-11eb-8011-2fdf3b6446ea.png">
<img width="1515" alt="Screen Shot 2020-12-09 at 2 49 00 PM" src="https://user-images.githubusercontent.com/1271364/101603958-b08efe80-3a2e-11eb-8263-fc92030b9446.png">
<img width="1521" alt="Screen Shot 2020-12-09 at 2 48 21 PM" src="https://user-images.githubusercontent.com/1271364/101603960-b1279500-3a2e-11eb-8282-b2600bf621d0.png">

kevcodez commented:
@jennifer-shehane Thank you very much, I'll have a look into the differences and will get back to you when I am done.

Just a suggestion: It would be great if contributors had any way of running those percy tests locally. That way, I wouldn't have to bother you with executing the tests. 👍

kevcodez commented:
@jennifer-shehane Could you please rerun the Percy build and send me the link again?

* Fixed button styling
* Changed base font size from 16px to 14px (0.875rem) to be compatible with bootstrap 3
* Changed base font family to be compatible with bootstrap 3
* Changed base line height to be compatible with bootstrap 3
* Changed modal width to align with the previous modal
* Fixed styling of several components

jennifer-shehane commented:
@kevcodez Yah, I also wish they could be run but we discussed it and did not want to expose our API key.

I've initiated a new Percy build: https://percy.io/cypress-io/cypress/builds/8036142

Saibamen commented on packages/desktop-gui/src/styles/components/_alerts.scss:
```diff
@@ -129,3 +129,23 @@
+}
+
+.alert.alert-info {
+	color: #31708f;
```
Identation

jennifer-shehane commented:
There's still some desktop-gui tests that are failing.

kevcodez commented:
@jennifer-shehane Fixed the failing tests, the only one still failing is *e2e record parallelization passes in parallel with group - passes in parallel with group* in the server-e2e-tests-electron.

It will still take me some time to fix the remaining styling issues as I have little time right now, just FYI.

Saibamen commented on packages/desktop-gui/src/styles/components/_alerts.scss:
```diff
@@ -139,7 +139,7 @@
 
 .alert.alert-info {
-	color: #31708f;
+	  color: #31708f;
```
```suggestion
    color: #31708f;
```

kevcodez commented:
As discussed with @JessicaSachs on Discord - I'll directly upgrade to Bootstrap 5.

As of now, it is still in Beta. However, this would be a future-proof move. Otherwise, the framework would likely be outdated by an entire major version shortly after merging.

jennifer-shehane commented:
@kevcodez Sure, just ping us if you need the Percy snapshots rerun at any time.

kevcodez commented:
@jennifer-shehane Could you please re-run? I know there will be many changes, especially with the BS5 upgrade  - just want to see if some UI is completely off.

I don't think it makes any sense to try to get the UI 100% as in Bootstrap 3. That would require many CSS adjustments. There's already so much CSS scattered all of the place. When I fixed the major remaining styling issues, I'd like to have a chat with the designer to verify some smaller changes, I'll contact someone on Discord.

jennifer-shehane commented:
Sorry, this PR is lost in my notifications.

@kevcodez 
- https://percy.io/cypress-io/cypress/builds/9121410
- https://percy.io/cypress-io/cypress/builds/9123058
- https://percy.io/cypress-io/cypress/builds/9127676

jennifer-shehane commented:
@kevcodez Will you have time to address the rest of the changes in the PR. This is probably a bit of a larger PR than our team originally had time to see through every step of the way unfortunately.

kevcodez commented:
@jennifer-shehane Unfortunately I don't have the time to continue this change as of now. Anyone who feels free to build up on this, please go ahead.

Not sure if a general overhaul of the CSS (starting on a clean base) would be smarter. There is so much custom CSS in this project that overwrites Bootstrap styles making it very difficult to migrate.

jennifer-shehane commented:
Closing for now. We can reopen if we decide to reprioritize this work
